### ISSUE TITLE
How to register an existing hub with a different DI engine

### ISSUE THREAD

**Issue by grahambunce** [aspnet/Docs#7284]
How to I get an reference to an existing Hub created by the .UseSignalR route mapping and then register it with my existing DI container (e.g. Unity) so that I can inject it into my business logic layer?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7bf6fefc-557e-114a-a37c-90939e06b083
* Version Independent ID: 92da23e0-6dc5-170e-b489-2ad8bbd46dff
* Content: [SignalR HubContext](https://docs.microsoft.com/en-us/aspnet/core/signalr/hubcontext?view=aspnetcore-2.1)
* Content Source: [aspnetcore/signalr/hubcontext.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/signalr/hubcontext.md)
* Product: **aspnet-core**
* GitHub Login: @rachelappel
* Microsoft Alias: **rachelap**

**Comment by scottaddie** [comment]
@grahambunce Here's info on replacing the native IoC container with a different one: https://docs.microsoft.com/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-2.1#replacing-the-default-services-container. The example used here is Autofac. Does this help you solve the problem?

**Comment by grahambunce** [comment]
To update my own post, I've worked this out. In the .net core lifecycle the unity registrations are run before the signalr framework runs, hence why unity doesn't contain the hub.

Then, to achieve my goal of not having signalr interfaces / packages in my domain layer, I needed to create an adapter class in the api layer that took an injected IHubContext<Station>. This then allowed me to seperate my business logic from SignalR.

As expected, DI "just worked" once I had this in place as the IHubContext was added to unity by SignalR automatically, and my unity registrations could work on my own IRemoteStation interface and the RemoteStation adapter.
